src/main/java/shef/servlets/FollowersServlet.java

```diff
@@ -0,0 +1,92 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shef.servlets;
+
+import shef.data.User;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.LinkedList;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+
+/** Servlet that retrieves follower information. */
+@WebServlet("/followers")
+public class FollowersServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Get the key from the query string.
+    String userKeyString = request.getParameter("key");
+    Key userKey;
+    if(userKeyString == null && userService.isUserLoggedIn()) {
+      userKey = KeyFactory.createKey("User", userService.getCurrentUser().getUserId());
+    } else {
+      userKey = KeyFactory.stringToKey(userKeyString);
+    }
+
+    List<String> followers;
+    try {
+      Entity userEntity = datastore.get(userKey);
+      followers = (List<String>) userEntity.getProperty("followers");
+    } catch(EntityNotFoundException e) {
+      // User doesn't exist.
+      e.printStackTrace();
+      return;      
+    }
+
+
+    LinkedList<User> followerUsers = new LinkedList<User>();
+
+    if(followers != null) {
+      for(String keyString : followers) {
+        try {
+          Key key = KeyFactory.stringToKey(keyString);
+          Entity userEntity = datastore.get(key);
+          String name = (String) userEntity.getProperty("username");
+          String profilePicKey = (String) userEntity.getProperty("profile-pic");
+
+          User user = new User(keyString, name, profilePicKey);
+          followerUsers.add(user);
+        } catch(EntityNotFoundException e) {
+          e.printStackTrace();
```

Comment:
If a user's key isn't found (maybe the user's account was deleted or something), then this try-catch will completely interrupt the servlet and return no followers to the user. I think this behavior can be improved with a continue statement, so the problematic user is skipped, but the loop still gets all the other value users.

Reply:
I think it also might be better to run a query on Datastore users as opposed to iterating through a loop of keys and getting user data (I think individual calls to separate APIs are costly).